### DISCUSSION TITLE
handleSubmit does not update the payload when saving inputs have defaultValues

### DISCUSSION THREAD

**Discussion by Cants820** [react-hook-form/react-hook-form#8122] [Q&A]
Hello,

I was wondering if there is a solution for the case im trying to do. Basically I have a form that loads values from useState. What happens is when you save a form without interacting with any of the inputs. The defaultValues are not being passed in the handleSubmit.  Any help would be greatly appreciated.  Basically I have defaultValues shown on onload and saving it the payload without touching the form. The payload becomes empty I think getting the from the redux state.

Here is my Input: 

```
import React from 'react'
import styled from 'styled-components';


const Label = styled.label`
  font-weight: 500;
  width: 100%;
`;

const StyledInput = styled.input`
  text-indent: 3px;
  height: 25px !important;
  
  border: 1px solid
    ${props => (!props.error ? props.theme.colors.blue01 : props.theme.colors.errorRed)} !important;
  padding: 10px;
  `;

const Input = ({ onChangeHandler,type, name,label, register, required, disabled, defaultValue, maxLength, minLength, valueAsNumber }) => (
    <>
      <Label htmlFor={name}>{label}</Label>
      <StyledInput type={type ? type: 'text'} name={name} defaultValue={defaultValue} disabled={disabled} {...register(name, { required: required ? 'Required' : '', maxLength, minLength, valueAsNumber: valueAsNumber ? true : false, onChange: onChangeHandler  })} />
    </>
  );

export default Input;
```

Here is my onSubmit handler:

```
    const onSubmit = payload => {
        dispatch(saveDangerous([payload]));
    }

```

**Comment by Moshyfawn** [comment]
How do you get your `defaultValues`? You said `useState`, so I assume the data is async from some backend or whatnot. So, you probably want to reset your form with the `useState` data values once you get them inside the `useEffect` as a side-effect
```js
const [data, setData] = useState({})

const { reset } = useForm({
  defaultValues: data
})
// INFO: assuming the data is gonna be set async after some time
useEffect(() => {
  if (data) {
    reset(data)
  }
}, [reset, data])
```

From [the RHF docs on `useForm`](https://react-hook-form.com/api/useform):
> `defaultValues` are cached on the first render within the custom hook. If you want to [reset](https://react-hook-form.com/api/useform/reset) the defaultValues, you should use the reset api.

**Comment by ritikbanger** [comment]
I am unable to understand your case, you are not using the `FormProvider` or `handleSubmit`.

Consider this example to set default values:
 ```
const initialValues: <Interface> = {
    title: title,
  }

  const methods = useForm({
    defaultValues: useMemo(() => {
      return initialValues
    }, [initialValues]),
    resolver: yupResolver(validationSchema(translation)),
  })

  const {
    handleSubmit,
    register,
    reset,
    control,
    formState: { errors },
  } = methods
```

**Reply by Moshyfawn** [reply]
> const methods = useForm({
>    defaultValues: useMemo(() => {
>      return initialValues
>    }, [initialValues]),
>    resolver: yupResolver(validationSchema(translation)),
>  })

Here's [an example CSB](https://codesandbox.io/s/vigilant-bartik-mq86le?file=%2Fsrc%2FApp.tsx) I've managed to create from your description. It works as intended.

Please, consider tweaking it to better match your use-case.

One note: you don't need to memoize your `defaultValues`, as they're already cached on first render

[From RHF docs on `useForm`](https://react-hook-form.com/api/useform):
> `defaultValues` are cached on the first render within the custom hook. If you want to [reset](https://react-hook-form.com/api/useform/reset) the `defaultValues`, you should use the reset api.:
